add GradleRunIntegrationTest (incl. QuarkusJavaLauncher)

vorburger opened the pull request:
This fixes #1623

It uses https://github.com/vorburger/ch.vorburger.exec#advantages

vorburger committed:
add GradleRunIntegrationTest (fixes #1623)

vorburger committed:
use exec util instead of raw ProcessBuilder in GradleRunIntegrationTest

see https://github.com/vorburger/ch.vorburger.exec#advantages

vorburger commented:
Reviewers: Look [only at the last commit of this PR](https://github.com/quarkusio/quarkus/pull/1986/commits/00d389b3f013cfb54400f5cec2acefce50d63aa6) - everything before is #1985 & Co.

vorburger commented:
This may actually help to understand the cause of the build failure in #1985 better (because {my} exec util logs much more than j.l.Process), so unless there are strong objections to the inclusion of that 3rd party lib, I think this provides real value.

I'll therefore ASAP rebase this based on recent merge (of #1993 and #1929) and lets see if we get more details about the failure (only) on Azure...

stalep commented:
I'm just not sure the reason to add another dependency for this (which contain other dependencies :) other than just for logging. What kind of output is omitted when using standard Process vs your impl?

vorburger commented:
> What kind of output is omitted when using standard Process vs your impl?

There simply is no output at all... :smiling_imp: it would, of course, be possible to re-implement the equivalent, but (to me) this seems to belong in some library - which is, years ago, I built https://github.com/vorburger/ch.vorburger.exec, and am happy to also offer it to Quarkus... :smile_cat: See also https://github.com/vorburger/ch.vorburger.exec#advantages. (BTW: I'm happy to "donate" it under some JBoss.org repo, if there is interest, or accept other committers to the project.)

> help to understand the cause of the build failure in #1985 better 
> see if we get more details about the failure (only) on Azure with this...

And indeed it does!! FTR, locally it's like this:

```
[INFO] Running io.quarkus.gradle.it.GradleRunIntegrationTest
[main] INFO ch.vorburger.exec.ManagedProcess - Starting Program [java, -jar, ../gradle-it/build/gradle-it-runner.jar] (in working directory /home/vorburger/Quarkus/git/quarkus/devtools/gradle-it-run/.)
[Exec Stream Pumper] INFO ch.vorburger.exec.ManagedProcess - java: 2019-04-11 12:32:17,896 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT started in 3.055s. Listening on: http://[::]:8080
[Exec Stream Pumper] INFO ch.vorburger.exec.ManagedProcess - java: 2019-04-11 12:32:18,007 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy]
[main] INFO ch.vorburger.exec.ManagedProcess - Successfully destroyed Program [java, -jar, ../gradle-it/build/gradle-it-runner.jar] (in working directory /home/vorburger/Quarkus/git/quarkus/devtools/gradle-it-run/.)
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.591 s - in io.quarkus.gradle.it.GradleRunIntegrationTest
```

on the Azure build it's like this:

```
[INFO] Running io.quarkus.gradle.it.GradleRunIntegrationTest
[main] INFO ch.vorburger.exec.ManagedProcess - Starting Program [java, -jar, ../gradle-it/build/gradle-it-runner.jar] (in working directory /home/vsts/work/1/s/devtools/gradle-it-run/.)
[main] INFO ch.vorburger.exec.ManagedProcess - Successfully destroyed Program [java, -jar, ../gradle-it/build/gradle-it-runner.jar] (in working directory /home/vsts/work/1/s/devtools/gradle-it-run/.)
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.494 s <<< FAILURE! - in io.quarkus.gradle.it.GradleRunIntegrationTest
[ERROR] testHelloEndpoint  Time elapsed: 1.479 s  <<< ERROR!
java.net.ConnectException: Connection refused (Connection refused)
```

I now see the problem - and will ASAP amend this PR with what I think will fix it...

vorburger committed:
make GradleRunIntegrationTest wait for "Listening on: http"

This should fix the test failure on slow CI servers such as Azure Build.

vorburger commented:
The build failure (above) on the Build_JDK11_Linux Job is due to #2012 ... @stalep BTW this underlines how important it is to run tests like this and be able to see the output from the externally launched process. If this new `GradleRunIntegrationTest`, and more like it in the future (e.g. #2005) were to use only simple `j.l.ProcessBuilder` directly, it would be almost impossible to make sense of a test failure like the #2012 problem (trust me, I've been there, in MariaDB4j).

vorburger committed:
new QuarkusJavaLauncher which takes JAVA_HOME into account (fixes #2012)

This should fix the test failure on environments where the
wrong `java` is on the `PATH`, such as in Azure; for details, see
https://github.com/Microsoft/azure-pipelines-tasks/issues/10103.

stalep commented:
hm, the windows build failed again.

vorburger commented:
> hm, the windows build failed again

not just the Windows build but the Java 11 one as well :cry: see #2036, for some reason the built JAR is bad, it has `INFO [io.quarkus] (main) Installed features: []` instead of `[cdi, resteasy]` - but I've no clue what could cause this, on the Azure build; works fine locally!

geoand commented:
I also tried this PR locally with both Java 8 and Java 11 and it worked.

vorburger commented:
@stalep @cescoffier @aloubyansky  @gsmet @stuartwdouglas I'd love for this work to be able to be merged, but am quite clueless as to why the Gradle build produces a JAR with empty Installed Features - and thus fails this new IT, but only on the Azure build; it works locally, for me, @geoand and (I think) @stalep.

I'm happy to rebase this once - if someone can then help to investigate the build failure further?

stuartwdouglas commented on devtools/gradle-it-run/pom.xml:
```diff
@@ -0,0 +1,67 @@
+        <maven.deploy.skip>true</maven.deploy.skip>
+    </properties>
+
+    <dependencies>
```
This does not have a dependency on quarkus-gradle-plugin-integration-test so its possible this can run before the jar it depends on is built.

vorburger replied:
@stuartwdouglas that's a good point, I've added a dependency now - am curious if this helps to get this PR to finally pass on the Azure build server as well (and not just locally, as it always has)

cescoffier commented:
I'm a bit worried about this PR pending for so long. @stalep when you have a few minutes can you tell me the current state?

vorburger committed:
Revert "remove gradle-it, need a better approach"

This reverts commit 20ecd1b3b043954e58ed60449cd036c1de3c2ed6.

vorburger commented:
#2214 removed the `devtools/gradle-it` which this is based on, so that has to come back for this...

vorburger committed:
add src/main/resources to gradle-it to work around bug #2782

vorburger committed:
add explicit dependency from gradle-it-run/ to gradle-it/ to force build order

vorburger commented:
> #2214 removed the devtools/gradle-it which this is based on, so that has to come back for this...

taken care of in #2783

> cescoffier added the needs-rebase label 4 days ago

Rebase done (based on / assuming #2783, which is included here)

> tell me the current state?

It used to build fine locally and fail (only) on the Azure build; let's see if it's better now...

emmanuelbernard commented:
Ping @stalep

dmlloyd commented:
Ping x2 @stalep?  I'll fire off a more recent CI but it would be good to get a review on this.

dmlloyd commented:
Never mind: I don't have permissions to re-queue a successful build.  Also the build is so old that it has disappeared off Azure.  So I guess... maybe force-push it one more time?

geoand commented:
Is this superceded by #2783 ?

emmanuelbernard commented:
This one is likely very outdated. I'm going to close it. Open it again if I'm incorrect.
